File: Bloomberg/3867.sum-of-gcd-of-formed-pairs.go
package leetcode

//
// @lc app=leetcode id=3867 lang=golang
//
// [3867] Sum of GCD of Formed Pairs
//
// Notes
// Maintain the running maximum and append gcd(current value, running maximum)
// for each prefix. Sorting those derived values and pairing smallest with
// largest matches the Python strategy, then sum gcds of each pair. Time:
// O(n log n). Space: O(n).
//
// @lc code=start

import "sort"

func GcdSum3867(nums []int) int {
	prefixGCD := make([]int, 0, len(nums))
	runningMax := 0
	for _, value := range nums {
		if value > runningMax {
			runningMax = value
		}
		prefixGCD = append(prefixGCD, gcd3867(value, runningMax))
	}

	sort.Ints(prefixGCD)
	answer := 0
	left, right := 0, len(prefixGCD)-1
	for left < right {
		answer += gcd3867(prefixGCD[left], prefixGCD[right])
		left++
		right--
	}
	return answer
}

func gcd3867(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	if a < 0 {
		return -a
	}
	return a
}

// @lc code=end

File: pt-1/3867.sum-of-gcd-of-formed-pairs.java
/*
 * @lc app=leetcode id=3867 lang=java
 *
 * [3867] Sum of GCD of Formed Pairs
 *
 * Build the same transformed prefix values as the Python reference:
 * prefixGcd[i] = gcd(nums[i], max(nums[0..i])). Sort them, pair smallest with
 * largest, and sum gcds of each pair.
 *
 * Time: O(n log n). Space: O(n).
 */

import java.util.Arrays;

// @lc code=start
class Solution {
    public int gcdSum(int[] nums) {
        int[] prefixGcd = new int[nums.length];
        int runningMax = 0;
        for (int i = 0; i < nums.length; i++) {
            runningMax = Math.max(runningMax, nums[i]);
            prefixGcd[i] = gcd(nums[i], runningMax);
        }

        Arrays.sort(prefixGcd);
        int answer = 0;
        int left = 0;
        int right = prefixGcd.length - 1;
        while (left < right) {
            answer += gcd(prefixGcd[left++], prefixGcd[right--]);
        }
        return answer;
    }

    private int gcd(int a, int b) {
        while (b != 0) {
            int t = a % b;
            a = b;
            b = t;
        }
        return Math.abs(a);
    }
}
// @lc code=end
